src/lib.rs

```diff
@@ -65,27 +70,88 @@ impl Otel {
             }
         };
 
-        let (registry, meter_provider) = init_metrics(config);
+        let (registry, meter_provider) = init_metrics(config.clone());
+
+        // Gather a list of CA cert paths from logs and metrics config
+        let mut ca_cert_paths: HashSet<String> = HashSet::new();
+        if let Some(metric_targets) = config.metrics_export_targets {
+            for metric_target in metric_targets {
+                if let Some(ca_cert_path) = &metric_target.ca_cert_path {
+                    ca_cert_paths.insert(ca_cert_path.clone());
+                }
+            }
+        }
+        if let Some(log_targets) = config.log_export_targets {
+            for log_target in log_targets {
+                if let Some(ca_cert_path) = &log_target.ca_cert_path {
+                    ca_cert_paths.insert(ca_cert_path.clone());
+                }
+            }
+        }
+
+        let (shutdown_tx, shutdown_rx) = mpsc::channel(1);
+
         Otel {
             registry,
             meter_provider,
             logger_provider,
+            ca_cert_paths,
+            shutdown_tx,
+            shutdown_rx,
         }
     }
 
     /// Long running tasks for otel propagation.
-    pub async fn run(&self) {
-        if let Some(prometheus_registry) = &self.registry {
-            let _ = httpserver_init(
-                prometheus_registry.port,
-                prometheus_registry.registry.clone(),
-            )
-            .await;
+    ///
+    /// # Errors
+    /// * `OtelError::WatcherError` - If there is an error with the CA file watcher
+    pub async fn run(&mut self) -> Result<(), OtelError> {
+        // Create watcher task to detect changes to the CA files
+        let (mut ca_watcher, mut ca_watcher_rx) = create_watcher()?;
+        self.ca_cert_paths.iter().for_each(|path| {
+            let _ = ca_watcher
+                .watcher()
+                .watch(&PathBuf::from(path), RecursiveMode::NonRecursive);
+            debug!("watching for changes to {path}");
+        });
+
+        tokio::select! {
+            notification = ca_watcher_rx.recv(), if !self.ca_cert_paths.is_empty() => {
```

Comment:
curious - do we need the `if` here? IIUC, if `ca-certs` are empty, there will be nothing watched and the `rx.recv()` will never fire...

Reply:
Originally, I added it before I added the shutdown arm so that if there was no registry and no ca certs, run() would exit immediately as it did before my changes. Technically, now the `if` is not needed, however I think it's good to keep it because the docs say it will prevent the associated future from being polled. This seems more efficient (i.e. no wasted cycles polling unnecessarily).

Note: I added the shutdown arm so that we have a graceful shutdown scenario that returns Ok(()), rather than the other scenarios that both return Err.